lib/postprocessing/image.js

```diff
@@ -94,113 +93,50 @@ function processBasicImageOperations(img, instructions) {
         img.units('pixelsperinch');
         if (typeof instructions.convertToDpi === 'object') {
             const { xdpi, ydpi } = instructions.convertToDpi;
-            img = img.resample(xdpi, ydpi);
+            img.resample(xdpi, ydpi);
         } else {
-            img = img.resample(instructions.convertToDpi, instructions.convertToDpi);
+            img.resample(instructions.convertToDpi, instructions.convertToDpi);
         }
     }
-
-    console.log('Processing rendition... Done!');
-    return img;
 }
 
-// Watermarking
-function processWatermark(img, watermarkInstructions) {
-    console.log('Processing watermark...');
-    const placement = watermarkInstructions.placement || 'Center';
-    const visibilityPercentage = watermarkInstructions.visibilityPercentage || 100;
-
-    let watermark = watermarkInstructions.watermarkContent;
-    if(watermarkInstructions.isEmbedded){
-        // watermark embedded as base 64
-        console.log('Using embedded watermark');
-        watermark = 'inline:'+watermarkInstructions.watermarkContent;
-    }
-
-    img.dissolve(visibilityPercentage).gravity(placement);
-    img.composite(watermark);
+async function checkJpegSize(outfile, instructions) {
+    // If we specified jpegSize for JPEG output we need to verify that the size does not exceed that
+    if (instructions.fmt === "jpg" && instructions.jpegSize) {
 
-    console.log('Processing watermark... Done!');
-    return img;
-}
+        const stat = await fs.stat(outfile);
+        if (stat.size > instructions.jpegSize) {
+            const img = gm(outfile);
 
-function applyImageOperations(img, instructions, operations){
-    if(operations.basicProcessing) {
-        img = processBasicImageOperations(img, instructions);
-    }
+            // From http://www.imagemagick.org/Usage/formats/#jpg
+            // As of IM v6.5.8-2 you can specify a maximum output filesize for the JPEG image.
+            img.define(`jpeg:extent=${instructions.jpegSize}`);
 
-    if(operations.watermarking) {
-        if(instructions.watermark && typeof instructions.watermark === 'object'){
-            img = processWatermark(img, instructions.watermark);
+            await img.write(outfile);
```

Comment:
Why are you not using the `write` function defined below?

Reply:
It might be hard to read in the diff: our `write()` function here calls the `checkJpegSize()` that this line of code is in. If we would reuse it there might be a recursion risk. There wouldn't be much reuse benefit since we only need the `img.write()` here.

Reply:
I saw it. What I am saying is we could split that up a bit more to reuse write, and build a "wrapper" function calling the things in the proper order.

Reply:
Got it - did that here: https://github.com/adobe/asset-compute-sdk/pull/22/commits/d44fe4af5d011f217d0e485ec63e8d55ef0d52a7#diff-c8cfb029f1e204a8f6da461ffa1464b0R103-R113

(now the command is logged for that possible 2nd execution as well)

Reply:
Ok, and I [removed](https://github.com/adobe/asset-compute-sdk/pull/22/commits/c5ec37f6a9f23a90191b4244ee6d740b28c9ab6c) the need for the 2nd execution, and the code became simpler (make sure to scroll down in that code snippet): https://github.com/adobe/asset-compute-sdk/blob/c5ec37f6a9f23a90191b4244ee6d740b28c9ab6c/lib/postprocessing/image.js#L110-L133